### PR TITLE
Preserve SrcSpan during compilation

### LINKED ISSUE #34: Preserve source span information during the compilation
## Description
Prior to the switch to our own AST data structure, all source span information contained in the parsing result was explicitely discarded by replacing it with the unit type `()` before applying the pattern matching compiler. While this information is no longer thrown away generally at the start, it is still not retained during the compilation.
Instead of passing down the source span information, `()` was inserted at most places in the code, which has now simply been replaced by `S.NoSrcSpan`. The task is to minimize the usage of this constructor as much as possible and instead pass down the original source span if available. Whenever new AST constructs are added or existing constructs are converted to other AST constructs during the pattern matching compilation, the best fitting source span should be inserted, if possible.

## Motivation
If some error occured during the pattern matching compilation, preserving the source span information allows printing out the exact place in the input program where something went wrong. Additionally, we can even pass down these data if the output of the pattern matching compiler is processed further, for example by the [Free Compiler](https://github.com/FreeProving/free-compiler), so that these programs can print better error messages as well.

### PR DESCRIPTION
**Pull request by stu204146-kiel**
### Issue

The `SrcSpan`s are preserved as much as possible for better error messages.
Closes #34.

### Description of the Change
I have changed the code of the features such that the source-span information is preserved during the features.

### Verification Process

I tried to test several programs over which I ran the compiler.

### Additional Notes
Maybe guard elimination contains some bugs.
I have added a `SrcSpan`-attribute to `GExp`.

### CODE REVIEW

File: src/lib/HST/Application.hs [hunk 1]
```diff
@@ -83,13 +83,14 @@ useAlgo :: (Members '[Env a, Fresh, GetOpt, Report] r, S.EqAST a)
...
-      arity = length (fst (head eqs))
+  let name    = getMatchName (head ms)
+      arity   = length (fst (head eqs))
+      rhsSpan = S.getSrcSpan (snd (head eqs))
   nVars <- replicateM arity (freshVarPat genericFreshPrefix)
   nExp <- match nVars eqs defaultErrorExp
   nExp' <- ifM (getOpt optOptimizeCase) (optimize nExp) (return nExp)
   return
-    $ S.Match S.NoSrcSpan name nVars (S.UnGuardedRhs S.NoSrcSpan nExp') Nothing
+    $ S.Match S.NoSrcSpan name nVars (S.UnGuardedRhs rhsSpan nExp') Nothing
```
Comment (stu204767) [inline src/lib/HST/Application.hs]: I think the source span of the corresponding `FunBind` should be used here, both for the `Match` and the `UnGuardedRhs`. This `rhsSpan` seems to belong to the first match of the original function only, while we want to span all matches.

File: src/lib/HST/CoreAlgorithm.hs [hunk 2]
```diff
@@ -194,9 +194,10 @@ createAltsForMissingCons x cs er = mapM (createAltForMissingCon x er) cs
   createAltForMissingCon pat e conEntry = do
     nvars <- replicateM (conEntryArity conEntry)
       (freshVarPat genericFreshPrefix)
-    let p    | conEntryIsInfix conEntry = S.PInfixApp S.NoSrcSpan (head nvars)
-               (conEntryName conEntry) (nvars !! 1)
-             | otherwise = S.PApp S.NoSrcSpan (conEntryName conEntry) nvars
+    let p    | conEntryIsInfix conEntry = S.PInfixApp (S.getSrcSpan pat)
+               (head nvars) (conEntryName conEntry) (nvars !! 1)
+             | otherwise = S.PApp (S.getSrcSpan pat) (conEntryName conEntry)
+               nvars
```
Comment (stu204767) [inline src/lib/HST/CoreAlgorithm.hs]: If I backtracked the code correctly, then this `S.Pat` is (always?) generated in the `freshVarPat` function in the `HST.Effect.Fresh` module, where it gets `S.NoSrcSpan` as its source span.
While I do think that there is no perfect source span for a generated case alternative, there probably exists a better solution, possibly using the equations (`Eqs`) that occur in most other functions of this module.

File: src/lib/HST/CoreAlgorithm.hs [hunk 3]
```diff
@@ -306,30 +307,29 @@ computeAlt _ _ _ []
...
-  | null ps = return (S.PList S.NoSrcSpan [], [], [])
+decomposeConPat (S.PList s ps)
+  | null ps = return (S.PList s [], [], [])
   | otherwise = do
     let (n : nv) = ps
+        nspan    = S.getSrcSpan n
         listCon  = S.Special S.NoSrcSpan $ S.ConsCon S.NoSrcSpan
-    decomposeConPat (S.PInfixApp S.NoSrcSpan n listCon (S.PList S.NoSrcSpan nv))
-decomposeConPat (S.PTuple _ bxd ps)         = do
+    decomposeConPat (S.PInfixApp nspan n listCon (S.PList S.NoSrcSpan nv))
```
Comment (stu204767) [inline src/lib/HST/CoreAlgorithm.hs]: I think we should use the source span of the entire list and not the one for the first element here. We don't have a good source span for the remaining list, so we should use the source span of the entire list again in my opinion.
```suggestion
    decomposeConPat (S.PInfixApp s n listCon (S.PList s nv))
```

File: src/lib/HST/Feature/CaseCompletion.hs [hunk 4]
```diff
@@ -74,26 +74,27 @@ completeCase _ e@(S.Lit _ _)              = return e
...
 
 completeLambda :: (Members '[Env a, Fresh, GetOpt, Report] r, S.EqAST a)
-               => [S.Pat a]
+               => S.SrcSpan a
+               -> [S.Pat a]
                -> S.Exp a
                -> Bool
                -> Sem r (S.Exp a)
-completeLambda ps e insideLet = do
+completeLambda s ps e insideLet = do
```
Comment (stu204767) [inline src/lib/HST/Feature/CaseCompletion.hs:92]: Nice, adding an additional parameter like in this case would help removing some of the remaining `NoSrcSpan`s too.

File: src/lib/HST/Feature/GuardElimination.hs [hunk 5]
```diff
@@ -11,16 +11,19 @@ import qualified HST.Frontend.Syntax as S
 
 -- | A pair of patterns to match and a right-hand side to use when all
 --   patterns match.
-data GExp a = GExp { gExpPats :: [S.Pat a], gExpRhs :: S.Rhs a }
+data GExp a = GExp { gExpSrcSpan :: S.SrcSpan a
+                   , gExpPats    :: [S.Pat a]
+                   , gExpRhs     :: S.Rhs a
+                   }
```
Comment (stu204767) [inline src/lib/HST/Feature/GuardElimination.hs:18]: Nice.

File: src/lib/HST/Feature/GuardElimination.hs [hunk 6]
```diff
@@ -194,14 +197,14 @@ applyGEMatches ms | any hasGuards ms = return <$> applyGE ms
...
+  let name    = getMatchName (head ms)
+      lhsSpan = S.getSrcSpan (head ms)
+      gexps   = map matchToGExp ms
+      arity   = length (gExpPats (head gexps))
   varPats <- replicateM arity (freshVarPat genericFreshPrefix)
   expr' <- generateLet (map S.patToExp varPats) defaultErrorExp gexps
   return
-    $ S.Match S.NoSrcSpan name varPats (S.UnGuardedRhs S.NoSrcSpan expr')
-    Nothing
+    $ S.Match lhsSpan name varPats (S.UnGuardedRhs S.NoSrcSpan expr') Nothing
```
Comment (stu204767) [inline src/lib/HST/Feature/GuardElimination.hs]: Just as in the `useAlgo` function in the `Application` module, you could pass the source span of the `FunBind` to this function and use it here for the `Match` and the `UnGuardedRhs`.

**Review by stu204767 — CHANGES_REQUESTED** [review]
There are some more `S.NoSrcSpan`s left that probably can be easily avoided.
Also, you apparently haven't yet looked at some of the other modules containing unnecessary `S.NoSrcSpan`s, especially `HST.Feature.Optimization`.

File: src/lib/HST/Effect/Fresh.hs [hunk 7]
```diff
@@ -40,18 +44,41 @@ makeSem ''Fresh
 freshName :: Member Fresh r => String -> Sem r (S.Name a)
 freshName prefix = S.Ident S.NoSrcSpan <$> freshIdent prefix
```
Comment (just95) [inline src/lib/HST/Effect/Fresh.hs]: You could express the old functions in terms of their new `*WithSpan` variations.

```suggestion
freshName = flip freshNameWithSpan S.NoSrcSpan
```

I would also consider to call them `*WithSrcSpan` for consistency.

File: src/lib/HST/Feature/Optimization.hs [hunk 8]
```diff
@@ -152,12 +153,14 @@ renameAll ((from, to) : r) e = do
 --   While an alternative is optimized, the pattern is pushed to the stack
 --   of matched patterns for the scrutinee in the environment.
 addAndOpt :: Members '[PatternStack a, Fresh, Report] r
-          => S.QName a
+          => S.SrcSpan a
+          -> S.SrcSpan a
```
Comment (just95) [inline src/lib/HST/Feature/Optimization.hs]: Add documentation to these parameters to make clear which source span is which.

**Review by just95 — CHANGES_REQUESTED** [review]

**Review by just95 — APPROVED** [review]

File: src/lib/HST/Effect/Fresh.hs [hunk 9]
```diff
@@ -38,19 +42,43 @@ makeSem ''Fresh
 
 -- | Generates a name for a fresh variable.
 freshName :: Member Fresh r => String -> Sem r (S.Name a)
-freshName prefix = S.Ident S.NoSrcSpan <$> freshIdent prefix
+freshName = flip freshNameWithSrcSpan S.NoSrcSpan
+
+-- | Generates a name for a fresh variable and sets the `S.SrcSpan` to the given
+--   value.
```
Comment (stu204767) [inline src/lib/HST/Effect/Fresh.hs]:
```suggestion
-- | Generates a name for a fresh variable and sets the 'S.SrcSpan' to the
--   given value.
```

File: src/lib/HST/Effect/Fresh.hs [hunk 10]
```diff
@@ -38,19 +42,43 @@ makeSem ''Fresh
...
+  :: Member Fresh r => String -> S.SrcSpan a -> Sem r (S.Name a)
+freshNameWithSrcSpan prefix s = S.Ident s <$> freshIdent prefix
 
 -- | Generates an unqualified name for a fresh variable.
 freshQName :: Member Fresh r => String -> Sem r (S.QName a)
-freshQName prefix = S.UnQual S.NoSrcSpan <$> freshName prefix
+freshQName = flip freshQNameWithSrcSpan S.NoSrcSpan
+
+-- | Generates an unqualified name for a fresh variable and sets the `S.SrcSpan`
+--   to the given value.
```
Comment (stu204767) [inline src/lib/HST/Effect/Fresh.hs]:
```suggestion
-- | Generates an unqualified name for a fresh variable and sets the
--   'S.SrcSpan' to the given value.
```

File: src/lib/HST/Effect/Fresh.hs [hunk 11]
```diff
@@ -38,19 +42,43 @@ makeSem ''Fresh
...
+  :: Member Fresh r => String -> S.SrcSpan a -> Sem r (S.QName a)
+freshQNameWithSrcSpan prefix s = S.UnQual s <$> freshNameWithSrcSpan prefix s
 
 -- | Generates a fresh variable expression.
 freshVar :: Member Fresh r => String -> Sem r (S.Exp a)
-freshVar prefix = S.Var S.NoSrcSpan <$> freshQName prefix
+freshVar = flip freshVarWithSrcSpan S.NoSrcSpan
+
+-- | Generates a fresh variable expression and sets the `S.SrcSpan` to the given
+--   value.
```
Comment (stu204767) [inline src/lib/HST/Effect/Fresh.hs]:
```suggestion
-- | Generates a fresh variable expression and sets the 'S.SrcSpan' to the
--   given value.
```

File: src/lib/HST/Effect/Fresh.hs [hunk 12]
```diff
@@ -38,19 +42,43 @@ makeSem ''Fresh
...
+freshVarWithSrcSpan
+  :: Member Fresh r => String -> S.SrcSpan a -> Sem r (S.Exp a)
+freshVarWithSrcSpan prefix s = S.Var s <$> freshQNameWithSrcSpan prefix s
 
 -- | Generates a fresh variable pattern.
 freshVarPat :: Member Fresh r => String -> Sem r (S.Pat a)
-freshVarPat prefix = S.PVar S.NoSrcSpan <$> freshName prefix
+freshVarPat = flip freshVarPatWithSrcSpan S.NoSrcSpan
+
+-- | Generates a fresh variable pattern which `S.SrcSpan` is identical to the
```
Comment (stu204767) [inline src/lib/HST/Effect/Fresh.hs]:
```suggestion
-- | Generates a fresh variable pattern whose 'S.SrcSpan' is identical to the
```

File: src/lib/HST/Feature/Optimization.hs [hunk 13]
```diff
@@ -152,12 +153,14 @@ renameAll ((from, to) : r) e = do
 --   While an alternative is optimized, the pattern is pushed to the stack
 --   of matched patterns for the scrutinee in the environment.
 addAndOpt :: Members '[PatternStack a, Fresh, Report] r
-          => S.QName a
+          => S.SrcSpan a  -- ^ The source span for the whole case-expression
+          -> S.SrcSpan a  -- ^ The source span for the matched expression
```
Comment (stu204767) [inline src/lib/HST/Feature/Optimization.hs]:
```suggestion
          => S.SrcSpan a  -- ^ The source span for the whole case expression.
          -> S.SrcSpan a  -- ^ The source span for the matched expression.
```

**Review by stu204767 — CHANGES_REQUESTED** [review]
There are still three suggestions left from my last review. Have you looked at them?

File: src/lib/HST/Application.hs [hunk 14]
```diff
@@ -84,10 +83,10 @@ useAlgo :: (Members '[Env a, Fresh, GetOpt, Report] r, S.EqAST a)
         -> Sem r (S.Match a)
 useAlgo s ms = do
   eqs <- mapM matchToEquation ms
-  let name    = S.matchName (head ms)
-      arity   = length (fst (head eqs))
-      isInfix = all S.matchIsInfix ms
-  nVars <- replicateM arity (freshVarPat genericFreshPrefix)
+  let name     = S.matchName (head ms)
+      srcSpans = (map S.getSrcSpan) . S.matchPats . head $ ms
```
Comment (stu204767) [inline src/lib/HST/Application.hs]: Shouldn't these parantheses be redundant?
```suggestion
      srcSpans = map S.getSrcSpan . S.matchPats . head $ ms
```

**Review by stu204767 — CHANGES_REQUESTED** [review]
Okay, this seems to address one of the remaining suggestions I had, but what about [this one](https://github.com/FreeProving/haskell-src-transformations/pull/46#discussion_r477304108)?
Otherwise, you still have to merge the `main` branch into this one.

**Review by stu204767 — APPROVED** [review]

**Pull request merged**